internal/kms/kmip.go

```diff
@@ -0,0 +1,512 @@
+	}
+
+	var respDiscoverVersionsPayload kmip.DiscoverVersionsResponsePayload
+	err = decoder.DecodeValue(&respDiscoverVersionsPayload, ttlvPayload)
+	if err != nil {
+		return err
+	}
+
+	if len(respDiscoverVersionsPayload.ProtocolVersion) != 1 {
+		return fmt.Errorf("invalid len of discovered protocol versions %v expected 1",
+			len(respDiscoverVersionsPayload.ProtocolVersion))
+	}
+	pv := respDiscoverVersionsPayload.ProtocolVersion[0]
+	if pv.ProtocolVersionMajor != protocolMajor || pv.ProtocolVersionMinor != protocolMinor {
+		return fmt.Errorf("invalid discovered protocol version %v.%v expected %v.%v",
+			pv.ProtocolVersionMajor, pv.ProtocolVersionMinor, protocolMajor, protocolMinor)
+	}
+
+	return nil
+}
+
+// send sends KMIP operation over tls connection, return response and error.
+func (kms *kmipKMS) send(
+	conn io.ReadWriter,
+	operation kmip14.Operation,
+	payload interface{},
+) (*kmip.ResponseMessage, *ttlv.Decoder, []byte, error) {
+	biID := uuid.New()
+
+	msg := kmip.RequestMessage{
+		RequestHeader: kmip.RequestHeader{
+			ProtocolVersion: kmip.ProtocolVersion{
+				ProtocolVersionMajor: protocolMajor,
+				ProtocolVersionMinor: protocolMinor,
+			},
+			BatchCount: 1,
+		},
+		BatchItem: []kmip.RequestBatchItem{
+			{
+				UniqueBatchItemID: biID[:],
+				Operation:         operation,
+				RequestPayload:    payload,
+			},
+		},
+	}
+
+	req, err := ttlv.Marshal(msg)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	_, err = conn.Write(req)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	decoder := ttlv.NewDecoder(bufio.NewReader(conn))
+	resp, err := decoder.NextTTLV()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var respMsg kmip.ResponseMessage
+	err = decoder.DecodeValue(&respMsg, resp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return &respMsg, decoder, biID[:], nil
+}
+
+// Verify the response success and return the batch item.
+func (kms *kmipKMS) response(
+	respMsg *kmip.ResponseMessage,
+	operation kmip14.Operation,
+	uniqueBatchItemID []byte,
+) (*kmip.ResponseBatchItem, error) {
+	if respMsg.ResponseHeader.BatchCount != 1 {
+		return nil, fmt.Errorf("batch count %v should be 1", respMsg.ResponseHeader.BatchCount)
+	}
+	if len(respMsg.BatchItem) != 1 {
+		return nil, fmt.Errorf("batch Intems list len %v should be 1", len(respMsg.BatchItem))
+	}
+	bi := respMsg.BatchItem[0]
+	if operation != bi.Operation {
+		return nil, fmt.Errorf("unexpected operation, real %v expected %v", bi.Operation, operation)
+	}
+	if !bytes.Equal(uniqueBatchItemID, bi.UniqueBatchItemID) {
+		return nil, fmt.Errorf("unexpected uniqueBatchItemID, real %v expected %v", bi.UniqueBatchItemID, uniqueBatchItemID)
+	}
+	if kmip14.ResultStatusSuccess != bi.ResultStatus {
+		return nil, fmt.Errorf("unexpected result status %v expected success %v", bi.ResultStatus, kmip14.ResultStatusSuccess)
+	}
+
+	return &bi, nil
+}
+
+// TODO: use the following structs from https://github.com/gemalto/kmip-go
```

Comment:
is there any Issue opened in https://github.com/gemalto/kmip-go for tracking?

Reply:
https://github.com/ThalesGroup/kmip-go/issues/21